---
title: "ciber"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,message = F,warning = F)
```

ç›®å‰ä¸»æµçš„å…ç–«æµ¸æ¶¦è®¡ç®—æ–¹æ³•æ˜¯CIBERSORTå’Œssgseaï¼Œä»Šå¤©ä»‹ç»CIBERSORTã€‚

### 1.è¾“å…¥æ•°æ®è¦ä»€ä¹ˆ

ä¸‹é¢è¿™æ®µè¯æ‘˜è‡ªCIBERSORTçš„ä»‹ç»

> Importantly, all expression data should be non-negative, devoid of missing values, and represented in non-log linear space. 
>
>For Affymetrix microarrays, a custom chip definition file (CDF) is recommended (see Subheading 3.2.2) and should be normalized with MAS5 or RMA. 
>
>Illumina Beadchip and single color Agilent arrays should be processed as described in the limma package. 
>
>Standard RNA-Seq expression quantification metrics, such as frag- ments per kilobase per million (FPKM) and transcripts per kilobase million (TPM), are suitable for use with CIBERSORT.
--ã€ŠProfiling Tumor Infiltrating Immune Cells with CIBERSORTã€‹

éå¸¸æ¸…æ¥šçš„å†™å‡ºäº†è¾“å…¥æ•°æ®çš„è¦æ±‚ï¼š
1.ä¸å¯ä»¥æœ‰è´Ÿå€¼å’Œç¼ºå¤±å€¼
2.ä¸è¦å–log
3.å¦‚æœæ˜¯èŠ¯ç‰‡æ•°æ®ï¼Œæ˜‚é£èŠ¯ç‰‡ä½¿ç”¨RMAæ ‡å‡†åŒ–ï¼ŒIllumina çš„Beadchip å’ŒAgilentçš„å•è‰²èŠ¯ç‰‡ï¼Œç”¨limmaå¤„ç†ã€‚ 
4.å¦‚æœæ˜¯RNA-Seqè¡¨è¾¾é‡ï¼Œä½¿ç”¨FPKMå’ŒTPMéƒ½å¾ˆåˆé€‚ã€‚

èŠ¯ç‰‡çš„è¦æ±‚å¯èƒ½æŠŠä½ å”¬ä½äº†ï¼ŒGEOå¸¸è§„çš„è¡¨è¾¾çŸ©é˜µéƒ½æ˜¯è¿™æ ·å¾—åˆ°çš„ï¼Œç›´æ¥ä¸‹è½½ä½¿ç”¨å³å¯ã€‚æ³¨æ„æœ‰çš„è¡¨è¾¾çŸ©é˜µä¸‹è½½ä¸‹æ¥å°±å·²ç»å–è¿‡logï¼Œéœ€è¦é€†è½¬å›å»ã€‚æœ‰çš„ç»è¿‡äº†æ ‡å‡†åŒ–æˆ–è€…æœ‰è´Ÿå€¼ï¼Œéœ€è¦å¤„ç†åŸå§‹æ•°æ®ï¼Œå‰é¢å†™è¿‡ä»‹ç»æ–‡äº†ã€‚
  
### 3.æ¥ä¸€ä¸ªç¤ºä¾‹

#### 3.1.ä¸‹è½½TCGAçš„RNA-seqè¡¨è¾¾æ•°æ®

æœ‰å¤šä¸ªæ¸ é“å¯ä»¥ä¸‹è½½countæˆ–è€…fpkmæ•°æ®ã€‚å…¶å®fpkmè½¬tpmæ›´æ— ç—›ï¼Œä½†å› ä¸ºä¹‹å‰çš„æ•™ç¨‹éƒ½æ˜¯åªä¸‹è½½countï¼Œåšåç»­çš„å·®å¼‚åˆ†æï¼Œæˆ‘ä¹Ÿä¸æƒ³å†å›è¿‡å¤´å»ä¸‹è½½fpkmäº†ã€‚å°±åœ¨countåŸºç¡€ä¸Šè½¬tpmå³å¯ã€‚

å¾—åˆ°TCGA-CHOL_gdc.Rdataçš„æ–¹æ³•å¯å‚è€ƒï¼šTCGA-1.GDCæ•°æ®ä¸‹è½½

```{r}
rm(list = ls())
library(tinyarray)
library(tidyverse)
load("TCGA-CHOL_gdc.Rdata")
exp[1:4,1:4]
# è¡¨è¾¾çŸ©é˜µçš„è¡Œåè½¬æ¢æˆgenesymbol
exp = trans_exp(exp,mrna_only = T)
exp[1:4,1:4]
```

ä»countçŸ©é˜µå¾—åˆ°tpmï¼Œå‚è€ƒï¼šåŸºå› é•¿åº¦å¹¶ä¸æ˜¯end-startã€‚TCGAä½¿ç”¨çš„å‚è€ƒåŸºå› ç»„æ³¨é‡Šç‰ˆæœ¬æ˜¯genecodeV22ã€‚

#### 3.2.å°†countè½¬ä¸ºtpm

é¦–å…ˆæ˜¯è®¡ç®—åŸºå› æœ‰æ•ˆé•¿åº¦ï¼Œå› ä¸ºtcgaç»Ÿä¸€ä½¿ç”¨äº†v22ç‰ˆæœ¬ï¼Œæ‰€ä»¥æ›¿æ¢å…¶ä»–ç™Œç—‡å¹¶ä¸éœ€è¦é‡æ–°è®¡ç®—ï¼Œå¯ä»¥ç›´æ¥æ‹¿æ¥ç”¨çš„ã€‚

```{r}
if(F){
  library(rtracklayer)
  gtf = rtracklayer::import("gencode.v22.annotation.gtf.gz")
  class(gtf)
  gtf = as.data.frame(gtf);dim(gtf)
  table(gtf$type)
  exon = gtf[gtf$type=="exon",
           c("start","end","gene_name")]
  gle = lapply(split(exon,exon$gene_name),function(x){
    tmp=apply(x,1,function(y){
        y[1]:y[2]
    })
    length(unique(unlist(tmp)))
  })
  gle=data.frame(gene_name=names(gle),
               length=as.numeric(gle))
  save(gle,file = "v22_gle.Rdata")
}
load("v22_gle.Rdata")
head(gle)
```

åŸºå› é•¿åº¦éœ€è¦å’Œè¡¨è¾¾çŸ©é˜µè¡Œçš„é¡ºåºå¯¹åº”èµ·æ¥ï¼Œç”¨åˆ°Rè¯­è¨€åŸºç¡€é‡Œéå¸¸ä¼˜ç§€çš„ä¸€ä¸ªå‡½æ•°--matchã€‚

```{r}
le = gle[match(rownames(exp),gle$gene_name),"length"]

#è¿™ä¸ªå‡½æ•°æ˜¯ç°æˆçš„ã€‚
countToTpm <- function(counts, effLen)
{
    rate <- log(counts) - log(effLen)
    denom <- log(sum(exp(rate)))
    exp(rate - denom + log(1e6))
}

tpms <- apply(exp,2,countToTpm,le)
tpms[1:3,1:3]
```

è‡³æ­¤å¾—åˆ°äº†tpmçŸ©é˜µã€‚

#### 3.3 åšæˆcibersortè¦æ±‚çš„è¾“å…¥æ–‡ä»¶

è¿™ä¸ªç®—æ³•å¹¶æ²¡æœ‰è¢«å†™æˆRåŒ…ï¼Œè€Œæ˜¯åªæœ‰ä¸€ä¸ªæ”¾ç€å‡½æ•°çš„è„šæœ¬--CIBERSORT.Rï¼ŒæŠŠå®ƒä¸‹è½½ä¸‹æ¥æ”¾åœ¨å·¥ä½œç›®å½•å³å¯ã€‚

éœ€è¦ä¸¤ä¸ªè¾“å…¥æ–‡ä»¶:

ä¸€ä¸ªæ˜¯è¡¨è¾¾çŸ©é˜µæ–‡ä»¶

ä¸€ä¸ªæ˜¯å®˜ç½‘æä¾›çš„LM22.txtï¼Œè®°å½•äº†22ç§å…ç–«ç»†èƒçš„åŸºå› è¡¨è¾¾ç‰¹å¾æ•°æ®ã€‚

ç”±äºCIBERSORT.Rè¯»å–æ–‡ä»¶çš„ä»£ç æ¯”è¾ƒç²—æš´ï¼Œä¸ºäº†é€‚åº”å®ƒï¼Œå¯¼å‡ºæ–‡ä»¶ä¹‹å‰éœ€è¦æŠŠè¡Œåå˜æˆä¸€åˆ—ã€‚ä¸ç„¶åé¢å°±ä¼šæœ‰æŠ¥é”™ã€‚

```{r}
exp2 = as.data.frame(tpms)
exp2 = rownames_to_column(exp2)
write.table(exp2,file = "exp.txt",row.names = F,quote = F,sep = "\t")
```

#### 3.4. è¿è¡ŒCIBERSORT

```{r}
source("CIBERSORT.R")

if(F){
  TME.results = CIBERSORT("LM22.txt", 
                          "exp.txt" , 
                          perm = 1000, 
                          QN = T)
  save(TME.results,file = "ciber_CHOL.Rdata")
}
load("ciber_CHOL.Rdata")
TME.results[1:4,1:4]
re <- TME.results[,-(23:25)]
```

è¿è¡Œæœ‰äº›æ…¢ã€‚è®¡ç®—å‡ºæ¥çš„ç»“æœåŒ…å«äº†22ç§å…ç–«ç»†èƒçš„ä¸°åº¦ï¼Œè¿˜æœ‰ä¸‰åˆ—å…¶ä»–ç»Ÿè®¡é‡ï¼Œä¸ç®¡å®ƒä»¬ã€‚

#### 3.5. ç»å…¸çš„å…ç–«ç»†èƒä¸°åº¦çƒ­å›¾ 

é‚£äº›åœ¨ä¸€åŠä»¥ä¸Šæ ·æœ¬é‡Œä¸°åº¦ä¸º0çš„å…ç–«ç»†èƒï¼Œå°±ä¸å±•ç¤ºåœ¨çƒ­å›¾é‡Œäº†ã€‚æˆ‘çœ‹äº†ä¸€ä¸‹è¿™ä¸ªçƒ­å›¾ï¼Œä»èšç±»çš„æƒ…å†µæ¥çœ‹ï¼Œnormalå’Œtumoræ²¡æœ‰å¾ˆå¥½çš„åˆ†å¼€ã€‚

```{r}
library(pheatmap)
k <- apply(re,2,function(x) {sum(x == 0) < nrow(TME.results)/2})
table(k)
re2 <- as.data.frame(t(re[,k]))

an = data.frame(group = Group,
                row.names = colnames(exp))
pheatmap(re2,scale = "row",
         show_colnames = F,
         annotation_col = an,
         color = colorRampPalette(c("navy", "white", "firebrick3"))(50))

```


#### 3.6. ç›´æ–¹å›¾

å¯ä»¥å±•ç¤ºå‡ºæ¯ä¸ªæ ·æœ¬çš„å…ç–«ç»†èƒæ¯”ä¾‹

```{r}
library(RColorBrewer)
mypalette <- colorRampPalette(brewer.pal(8,"Set1"))

dat <- re %>% as.data.frame() %>%
  rownames_to_column("Sample") %>% 
  gather(key = Cell_type,value = Proportion,-Sample)

ggplot(dat,aes(Sample,Proportion,fill = Cell_type)) + 
  geom_bar(stat = "identity") +
  labs(fill = "Cell Type",x = "",y = "Estiamted Proportion") + 
  theme_bw() +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        legend.position = "bottom") + 
  scale_y_continuous(expand = c(0.01,0)) +
  scale_fill_manual(values = mypalette(22))
```

#### 3.7 ç®±çº¿å›¾

å±•ç¤ºå…ç–«ç»†èƒä¹‹é—´çš„æ¯”è¾ƒã€‚

```{r}
ggplot(dat,aes(Cell_type,Proportion,fill = Cell_type)) + 
  geom_boxplot(outlier.shape = 21,color = "black") + 
  theme_bw() + 
  labs(x = "Cell Type", y = "Estimated Proportion") +
    theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        legend.position = "bottom") + 
  scale_fill_manual(values = mypalette(22))
```

ä¹±äº†ç‚¹ï¼Ÿé‚£å°±è®©ç®±çº¿å›¾æ‹¥æœ‰é¡ºåºå§ã€‚

```{r}
a = dat %>% 
  group_by(Cell_type) %>% 
  summarise(m = median(Proportion)) %>% 
  arrange(desc(m)) %>% 
  pull(Cell_type)

dat$Cell_type = factor(dat$Cell_type,levels = a)

ggplot(dat,aes(Cell_type,Proportion,fill = Cell_type)) + 
  geom_boxplot(outlier.shape = 21,color = "black") + 
  theme_bw() + 
  labs(x = "Cell Type", y = "Estimated Proportion") +
    theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        legend.position = "bottom") + 
  scale_fill_manual(values = mypalette(22))
```

æ—¢ç„¶æˆ‘ä»¬å·²ç»æŠŠæ­£å¸¸æ ·æœ¬ä¹Ÿç®—äº†ï¼Œé‚£å°±åšä¸ªæ¯”è¾ƒï¼š

```{r}
dat$Group = ifelse(as.numeric(str_sub(dat$Sample,14,15))<10,"tumor","normal")
library(ggpubr)
ggplot(dat,aes(Cell_type,Proportion,fill = Group)) + 
  geom_boxplot(outlier.shape = 21,color = "black") + 
  theme_bw() + 
  labs(x = "Cell Type", y = "Estimated Proportion") +
  theme(legend.position = "top") + 
  theme(axis.text.x = element_text(angle=80,vjust = 0.5))+
  scale_fill_manual(values = mypalette(22)[c(6,1)])+ stat_compare_means(aes(group = Group,label = ..p.signif..),method = "kruskal.test")
```

åˆ†å¼€çœ‹çš„è¯ç¡®å®èƒ½çœ‹å‡ºåŒºåˆ«ğŸ‘ï¼Œåªæ˜¯ä¸æ˜¾è‘—çš„å¤ªå¤šäº†ã€‚
